internal/controller/consumer_controller.go

```diff
@@ -36,14 +45,264 @@ type ConsumerReconciler struct {
+	}
+
+	// Add finalizer
+	if !controllerutil.ContainsFinalizer(consumer, consumerFinalizer) {
+		log.Info("Adding consumer finalizer.")
+		if ok := controllerutil.AddFinalizer(consumer, consumerFinalizer); !ok {
+			return ctrl.Result{}, errors.New("failed to add finalizer to consumer resource")
+		}
+
+		if err := r.Update(ctx, consumer); err != nil {
+			return ctrl.Result{}, fmt.Errorf("update consumer resource to add finalizer: %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Check Deletion
+	markedForDeletion := consumer.GetDeletionTimestamp() != nil
+	if markedForDeletion {
+		if controllerutil.ContainsFinalizer(consumer, consumerFinalizer) {
+			err := r.deleteConsumer(ctx, log, consumer)
+			if err != nil {
+				return ctrl.Result{}, fmt.Errorf("delete consumer: %w", err)
+			}
+		} else {
+			log.Info("Consumer marked for deletion and already finalized. Ignoring.")
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// Create or update stream
+	if err := r.createOrUpdate(ctx, log, consumer); err != nil {
+		return ctrl.Result{}, fmt.Errorf("create or update: %s", err)
+	}
 	return ctrl.Result{}, nil
 }
 
+func (r *ConsumerReconciler) deleteConsumer(ctx context.Context, log logr.Logger, consumer *api.Consumer) error {
+
+	// Set status to not false
+	consumer.Status.Conditions = updateReadyCondition(consumer.Status.Conditions, v1.ConditionFalse, "Finalizing", "Performing finalizer operations.")
+	if err := r.Status().Update(ctx, consumer); err != nil {
+		return fmt.Errorf("update ready condition: %w", err)
+	}
+
+	if !consumer.Spec.PreventDelete && !r.ReadOnly() {
+		err := r.WithJetStreamClient(consumerConnOpts(consumer.Spec), func(js jetstream.JetStream) error {
+			return js.DeleteConsumer(ctx, consumer.Spec.StreamName, consumer.Spec.DurableName)
+		})
+		switch {
+		case errors.Is(err, jetstream.ErrConsumerNotFound):
+			log.Info("Consumer does not exist. Unable to delete.")
+		case errors.Is(err, jetstream.ErrStreamNotFound):
+			log.Info("Stream of consumer does not exist. Unable to delete.")
+		case err != nil:
+			return fmt.Errorf("delete jetstream consumer: %w", err)
+		default:
+			log.Info("Consumer deleted.")
+		}
+	} else {
+		log.Info("Skipping consumer deletion.",
+			"consumerName", consumer.Spec.DurableName,
+			"preventDelete", consumer.Spec.PreventDelete,
+			"read-only", r.ReadOnly(),
+		)
+	}
+
+	log.Info("Removing consumer finalizer.")
+	if ok := controllerutil.RemoveFinalizer(consumer, consumerFinalizer); !ok {
+		return errors.New("failed to remove consumer finalizer")
+	}
+	if err := r.Update(ctx, consumer); err != nil {
+		return fmt.Errorf("remove finalizer: %w", err)
+	}
+
+	return nil
+}
+
+func (r *ConsumerReconciler) createOrUpdate(ctx context.Context, log klog.Logger, consumer *api.Consumer) error {
+
+	// Create or Update the stream based on the spec
+	if consumer.Spec.PreventUpdate || r.ReadOnly() {
+		log.Info("Skipping consumer creation or update.",
+			"preventDelete", consumer.Spec.PreventDelete,
```

Comment:
Was this intended to log the `preventUpdate` setting in the spec rather than `preventDelete`?